### PR TITLE
[Model Runner V2] Support structured outputs + spec decoding

### PR DESCRIPTION
**Pull request by WoosukKwon**
> [!NOTE]
> Enables structured outputs to work with speculative decoding by introducing a dedicated worker and updating batching metadata.
> 
> - Adds `StructuredOutputsWorker` to apply grammar bitmasks in-place on logits using a new Triton kernel that maps bitmasks to specific logits via `logits_indices`
> - Integrates worker into `GPUModelRunner.sample`, replacing prior helper and supporting multiple logits per request (spec decoding)
> - Extends `InputBatch` with `cu_num_logits_np` and wires its population in `prepare_inputs` and dummy paths to support accurate logits mapping
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bdd44357294ad753ba3691a0de040637dc0f39a7. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces structured outputs compatibility with speculative decoding and updates batching metadata for accurate logits mapping.
> 
> - Adds `StructuredOutputsWorker` with a Triton kernel applying grammar bitmasks in-place to `logits` via `logits_indices`
> - Integrates worker into `GPUModelRunner.sample`, replacing prior helper and handling multiple logits per request
> - Extends `InputBatch` with `cu_num_logits_np`; populates it in `make_dummy` and `prepare_inputs` (both no-draft and draft paths)
> - Computes `cu_num_logits`/`expanded_idx_mapping` appropriately for spec decoding and uses them to build the bitmask→logits mapping
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 79e4cfc640165ab46d0cc6b9be16b3b99c4db428. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: vllm/v1/worker/gpu/structured_outputs.py [hunk 1]
```diff
@@ -4,38 +4,67 @@
...
+        req_id_to_cnt = {req_id: 0 for req_id in req_ids}
+        for grammar_req_id in grammar_req_ids:
+            req_idx = req_id_to_idx[grammar_req_id]
+            req_cnt = req_id_to_cnt[grammar_req_id]
+            logits_start_idx = cu_num_logits_np[req_idx]
+            mapping.append(logits_start_idx + req_cnt)
+            req_id_to_cnt[grammar_req_id] = req_cnt + 1
+        # Copy mapping to GPU
+        mapping_np = np.array(mapping, dtype=np.int32)
+        logits_indices = self.logits_indices.copy_to_gpu(mapping_np)
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/structured_outputs.py:51]: Good point. I misunderstood the input. Fixed.
